Integrate Adobe Stock search into search bar within DS

DivyaPrabhakar opened the pull request:
This PR adds the option to search Adobe Stock via the search bar. (JIRA PS-656) 
It also solves issues #3610, #3529, #3524. 

Testing: 

- The category "Adobe Stock" is now added to the search bar as a potential category the user can choose

- If the user selects the category "Adobe Stock", the user can: 
      > - Enter a string, press `enter` and adobestock.com will open to the results page of the user search 
      > - Enter no string, press `enter` and adobestock.com homepage will open 
      > - Enter a string, click the option `Search Adobe Stock...` and adobestock.com will open to the results page of the user search 
      > - Enter no string, click the option `Search Adobe Stock...` and adobestock.com homepage will open

- If the user selects any or no category within the search bar: 
      >  - If the user types in any string to the search bar and there are no results found, pressing `enter` should execute no action and nothing should happen to the state of the search bar.
      >  - If the user types in a string to the search bar and there are no results, and the user clicks the option `No Results Match`, nothing should happen to the state of the search bar. 
      >  - If the user presses `tab` to autocomplete and there is a a suggestion, the user inputted string should autocomplete. 
      >  - If the user presses `tab` to autocomplete and there is a a suggestion, the user inputted string should autocomplete. 
     >  - If the user presses `tab` to autocomplete and there is no suggestion, the user inputted string not change.
     > - If the user searches for a partial match of the option string, and then presses `enter`, the first option out of the list of options should select. Example: User types `documents`, the options list returns with "Open Documents", "Recent Documents", "documents-1353.psd", pressing enter should choose the option "Open Documents". 


- The category "Adobe Stock" is now added and after selecting it, entering a string within the input field and pressing "Enter" should open a new tab in the browser with the executed search and its results. 
- Searching Adobe Stock from DS will only happen when the "Adobe Stock" filter is applied. 
- When the "Adobe Stock" filter is applied, only the search bar is present, "No options for this result" should not be shown.
- From the browser, the user can choose an image and drag it directly into their document, save it to their local machine, or add it to their library.

iwehrman commented:
Let's wait to merge this until the "no results match" bug is fixed. I don't think we would want to ship this feature (even just in a sprint build) until that's resolved.

DivyaPrabhakar commented:
This PR has now been updated to reflect the bug fix, as per @iwehrman's comments. Ready for review.

iwehrman commented on src/js/actions/search.js:
```diff
@@ -59,11 +61,30 @@ define(function (require, exports) {
+     *
+     * @return {Promise}
+     */
+    var executeStockSearch = function (searchTerm) {
```
It seems like this should go in the `adobestock` module?

DivyaPrabhakar replied:
I was also debating that. The reason I included it in `actions/search.js` is because for every other search module, their `execute` option was in an actions module pertaining to the executed task. For example, the `executeMenuCommand()` function is held within `actions/menu.js`. Since searching Adobe Stock didn't have a corresponding actions module, I included it in the `search` module. I know it was a bit of a stretch but it follows suit for the other search modules. Having said that, with the second opinion, I think moving it to `adobesstock.js` makes sense and I'll make the change.

iwehrman commented on src/js/actions/search.js:
```diff
@@ -73,6 +94,22 @@ define(function (require, exports) {
+     * @return {Promise}
+     */
+    var afterStartup = function () {
+        return this.transfer("searchStock.registerAdobeStock");
```
And this too?

DivyaPrabhakar replied:
Done.

iwehrman commented on src/js/actions/search/adobestock.js:
```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2014 Adobe Systems Incorporated. All rights reserved.
```
The rule is: whenever you make a new file, make sure the copyright notice uses the current year. So, this should be 2016 instead of 2014.

Aside: for brevity, I probably would have just named this `stock.js`, but if you really like `adobestock.js` then OK!

iwehrman commented on src/js/actions/search/adobestock.js:
```diff
@@ -0,0 +1,86 @@
+     * @private
+     * @return {Immutable.List.<object>}
+    */
+    var _adobeStockSearchOptions = function () {
```
I don't understand why this is a method. `Immutable.List()` always returns the same value!

iwehrman replied:
Oh, I see what's happening now. But, I suggest just using an anonymous function literal in `registerAdobeStock` instead of a named, documented function as above. Also, it doesn't need to be bound with `.bind` because the body doesn't refer to `this`. (The only thing `.bind` does is set the value of `this`.)

DivyaPrabhakar replied:
The reasons I kept this as a function were: 

1.) In the case that we expand on the functionality of Adobe Stock search in the future, we will have options that will populate through this function. I saw this as an intermediate state which returns an empty list now, that is not unique now, but later on we will end up changing and wanting this function back.

2.) For every registered search payload, payload.getOptions is a function and previously when I had the `getOptions` property be an empty list instead of a function call, it would run into errors at different points in the program when `getOptions()` was called instead of `getOptions`. We can go through and have it check for function or value, etc. but a.) only the stock module requires this and b.) like I said above, I anticipate changing the return value of the function so we would change any calls to `getOptions` in the future back to the original `getOptions()`

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -45,7 +45,8 @@ define(function (require, exports, module) {
      */
-    var PLACEHOLDER_ID = "NO_OPTIONS-placeholder";
+    var PLACEHOLDER_ID = "NO_OPTIONS-placeholder",
+        ADOBE_STOCK_FILTER = "ADOBE_STOCK";
```
This is OK I guess, but it's kind of a bummer that you are directly tying one particular search provider into the search bar. It would be nicer if different search providers could optionally register "default" handlers. The default handler, in this case, would execute the stock search. It would also take the place of the "no results found" message. If I haven't explained that well, let's talk more!

DivyaPrabhakar replied:
I think I understand but I want to clarify in person with you before I go ahead with this.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -399,6 +400,13 @@ define(function (require, exports, module) {
+            var filters = this.state.filter,
+                value = event.target.value;
+
+            if (typeof value === "string" && filters.indexOf(ADOBE_STOCK_FILTER) !== -1) {
```
See above: it's lame that this click handler has stock-specific logic.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -416,10 +424,17 @@ define(function (require, exports, module) {
+                idSplit,
+                type;
+
+            if (id) {
```
This is weird. Above, we check for `id === null`. I'm not even sure what type `id` is supposed to have here because it's undocumented but I guess it can be some other falsey value like `undefined`? In any case, this smells funny.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -428,6 +443,7 @@ define(function (require, exports, module) {
                 filtersString = "filter-active";
             }
             
+
```
Nope!

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -463,7 +479,7 @@ define(function (require, exports, module) {
                 case "Return":
                 case "Enter":
-                    if (selectedID === PLACEHOLDER_ID) {
+                    if (selectedID === PLACEHOLDER_ID || !selectedID) {
```
If this disjunct is necessary, you should explain why in a comment. Also, these parameters are undocumented. JSDoc that explains what it means if `selectedID` is falsey would be a good idea.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -515,7 +531,7 @@ define(function (require, exports, module) {
                     categoryString = searchStrings.CATEGORIES[lastFilter],
                     category = categoryString ?
-                        categoryString.toLowerCase() : this.state.safeFilterNameMap[lastFilter];
+                        categoryString : this.state.safeFilterNameMap[lastFilter];
```
Was it not important before that `categoryString` be converted to lower case?

DivyaPrabhakar replied:
This conversion to lower case was inconsistent with how the strings in the search bar were being presented. 

In the dropdown, the options were capitalized.
![](https://www.dropbox.com/s/cnvlym5r8syy0hf/Screenshot%202016-01-06%2015.18.34.png?raw=1)

Upon selection, the placeholder text would become lower case. 
![](https://www.dropbox.com/s/airj8psgvzyw4io/Screenshot%202016-01-06%2015.18.56.png?raw=1)

I talked to Seth about this and he said removing the lower case upon selection makes sense.

iwehrman replied:
:white_check_mark:

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -543,6 +559,7 @@ define(function (require, exports, module) {
                         className="dialog-search-bar"
                         options={this.state.options}
                         startFocused={true}
+                        filters={filter}
```
It's weird that the singular word `filter` is being used to indicate a list of filters here?

iwehrman commented on src/js/jsx/shared/Datalist.jsx:
```diff
@@ -36,6 +36,8 @@ define(function (require, exports, module) {
         SVGIcon = require("js/jsx/shared/SVGIcon"),
         collection = require("js/util/collection");
 
+    var ADOBE_STOCK_FILTER = "ADOBE_STOCK";
```
Needs JSDoc.

iwehrman replied:
Nevermind, I just realized that this is in `Datalist`. Needs removal.

iwehrman commented on src/js/jsx/shared/Datalist.jsx:
```diff
@@ -314,6 +316,8 @@ define(function (require, exports, module) {
                 case "Tab":
                     return;
                 case "Enter":
+                    this.props.onKeyDown(event);
```
Uhh, why is `Enter` being handled differently from `Return` here?

DivyaPrabhakar replied:
I am not sure, I didn't even notice that. It also happens in `SearchBar.jsx` _handleKeyDown(). Removing the `Return` case from both.

iwehrman commented on src/js/jsx/shared/Datalist.jsx:
```diff
@@ -702,7 +706,11 @@ define(function (require, exports, module) {
+            var filters = this.props.filters;
+            var dialog;
+            
+            if (filters && filters.indexOf(ADOBE_STOCK_FILTER) === -1) {
```
It's definitely not acceptable for `Datalist` to have any knowledge of Adobe Stock. `Datalist` is a basic, shared component. This would be like Chrome adding Google-specific functionality to `<input type="text"/>`.

iwehrman commented on src/js/stores/search.js:
```diff
@@ -271,36 +281,41 @@ define(function (require, exports, module) {
+                        }
+
+                        if (name === "") {
+                            name = "Untitled";
```
I realize this isn't your code, but it sure seems weird that this string isn't localized. You might just want to file a separate bug for this?

iwehrman commented on src/js/stores/search.js:
```diff
@@ -271,36 +281,41 @@ define(function (require, exports, module) {
+                        var newPathInfo = shortenedPaths[index] || "",
+                            name = item.name,
+                            style;
+                        // Don't show the path info if it is just the same as the item name 
```
A newline should separate this declaration block from following statements.

iwehrman commented on src/js/actions/search.js:
```diff
@@ -59,11 +61,30 @@ define(function (require, exports) {
+     * @return {Promise}
+     */
+    var executeStockSearch = function (searchTerm) {
+        var searchURL = STOCK_URL + searchTerm;
```
You may need to URL-encode the search term: http://stackoverflow.com/a/3608791

iwehrman commented:
Back to @DivyaPrabhakar. Good start! I think it needs to be architected a little differently though so that the shared components don't have dependencies on Adobe Stock-specific code. I'm happy to chat about this offline!

iwehrman commented on src/js/jsx/shared/Datalist.jsx:
```diff
@@ -314,7 +316,8 @@ define(function (require, exports, module) {
                 case "Tab":
                     return;
                 case "Enter":
-                case "Return":
```
My guess is that you'll want to remove this code once you get the `defaultOptions` working because this code path is only active when the `Select` component is _not_ being shown.

iwehrman replied:
I still think you want to remove this code for the reason above.

iwehrman replied:
Also, the stock (or whatever default) result should work just like every other result. That means I should be able to click on it. But, I can't click on the stock result.

DivyaPrabhakar commented:
@iwehrman Back to you

iwehrman commented:
I opened the search UI and typed "s t o <enter>":
![img](https://s3.amazonaws.com/f.cl.ly/items/1K1k3e2H2y101t3x2B3u/Image%202016-01-13%20at%2010.08.14%20AM.png?v=e3322ef8)

and this happened:
![image](https://cloud.githubusercontent.com/assets/1075154/12303178/c462d4da-b9dd-11e5-9178-f79bdf70d8a0.png)

DivyaPrabhakar commented:
@iwehrman This error is actually happening on `master`. What is happening is that our string matching is currently only returning a successful result if the word entered matches the first word of the id of an option. For example: if I type in "sto", the list of results that are returned are "Adobe Stock". Then we go on to check if "sto" if positioned at the first index of the string "Adobe Stock". (see `datalist.jsx`, lines 556-568). Because it does not match, the autocomplete returns as `undefined` and then nothing gets stored as the state of the datalist `state` properites like `lastHighlightedID`. Later in the code, we use the value of `lastHighlightedID` to make decisions about executing datalist options on click. Because it is `undefined`,  we get the error that you see later when we are trying to split `id` that is `undefined`. 

If you type in "Adobe" and then select "Adobe Stock", then you will not run into any problem. We need to change our string matching -- should we do that in this PR?

iwehrman commented:
I get this warning in the console when I execute a search:
```
Warning: setState(...): Can only update a mounted or mounting component. This usually means you  called setState() on an unmounted component. This is a no-op. Please check the code for the TextInput component.

warning	@	warning.js:45
getInternalInstanceReadyForUpdate	@	ReactUpdateQueue.js:34
ReactUpdateQueue.enqueueSetState	@	ReactUpdateQueue.js:181
ReactComponent.setState	@	ReactComponent.js:65
(anonymous function)	@	TextInput.jsx:345
tryCatcher	@	bluebird.js:4978
Promise._settlePromiseFromHandler	@	bluebird.js:3042
Promise._settlePromise	@	bluebird.js:3099
Promise._settlePromise0	@	bluebird.js:3144
Promise._settlePromises	@	bluebird.js:3227
Promise._fulfill	@	bluebird.js:3168
Promise._settlePromise	@	bluebird.js:3112
Promise._settlePromise0	@	bluebird.js:3144
Promise._settlePromises	@	bluebird.js:3227
Promise._fulfill	@	bluebird.js:3168
(anonymous function)	@	bluebird.js:2482
```

iwehrman commented:
If I search for "100%" I get the following error:
```
Unhandled rejection Error: OpenURLInDefaltBrowser failed
    at eval (eval at undefined, <anonymous>:1:1)
    at eval (native)
From previous event:
    at Object.openURLInDefaultBrowser (file:///Users/wehrman/Desktop/photoshop-Release_x86_64/Plug-Ins/Spaces/www/build/spaces-design-en.js:415:23)
    at Object._confirmStockSearch (file:///Users/wehrman/Desktop/photoshop-Release_x86_64/Plug-Ins/Spaces/www/build/spaces-design-en.js:97398:25)
    at Fluxxor.createStore.handleExecute (file:///Users/wehrman/Desktop/photoshop-Release_x86_64/Plug-Ins/Spaces/www/build/spaces-design-en.js:74225:52)
    at null.<anonymous> (file:///Users/wehrman/Desktop/photoshop-Release_x86_64/Plug-Ins/Spaces/www/build/spaces-design-en.js:75579:34)
    at tryCatcher (file:///Users/wehrman/Desktop/photoshop-Release_x86_64/Plug-Ins/Spaces/www/build/externalDeps.js:5091:24)
    at Promise._settlePromiseFromHandler (file:///Users/wehrman/Desktop/photoshop-Release_x86_64/Plug-Ins/Spaces/www/build/externalDeps.js:3155:32)
    at Promise._settlePromise (file:///Users/wehrman/Desktop/photoshop-Release_x86_64/Plug-Ins/Spaces/www/build/externalDeps.js:3212:19)
    at Promise._settlePromise0 (file:///Users/wehrman/Desktop/photoshop-Release_x86_64/Plug-Ins/Spaces/www/build/externalDeps.js:3257:11)
    at Promise._settlePromises (file:///Users/wehrman/Desktop/photoshop-Release_x86_64/Plug-Ins/Spaces/www/build/externalDeps.js:3340:19)
    at file:///Users/wehrman/Desktop/photoshop-Release_x86_64/Plug-Ins/Spaces/www/build/externalDeps.js:275:26
    at MutationObserver.<anonymous> (file:///Users/wehrman/Desktop/photoshop-Release_x86_64/Plug-Ins/Spaces/www/build/externalDeps.js:4389:14)
```

iwehrman commented:
The above error probably has to do with URI encoding. I left a comment about this earlier.

iwehrman commented:
The message shown by the default option is currently too engineery. We don't need to explicitly instruct people to press enter. I think it should just say "Search Adobe Stock for "<your current search query>". The `Datalist` component should be able to provide the current search query to the `SearchBar`. Also, it should be disabled until you actually type something in; pressing enter in that case should have no effect.  Alternatively, the default option message could be different depending on whether the user has typed anything yet or not.

iwehrman commented on src/js/actions/search/stock.js:
```diff
@@ -0,0 +1,123 @@
+        var searchURL = STOCK_URL + searchTerm;
+        return adapter.openURLInDefaultBrowser(searchURL);
+    };
+    _confirmStockSearch.action = {
```
It doesn't look like this is an action; it's just a plain function. So, this property isn't needed.

iwehrman commented on src/js/actions/search/stock.js:
```diff
@@ -0,0 +1,123 @@
+     * @return {Promise}
+     */
+    var _confirmStockSearch = function (searchTerm) {
+        var searchURL = STOCK_URL + searchTerm;
```
This is what should be URI-encoded. See previous comment.

iwehrman commented on src/js/actions/search/stock.js:
```diff
@@ -0,0 +1,123 @@
+    var _createNoOptionsDefault = function () {
+        return {
+            "noOptionsString": nls.localize("strings.SEARCH.NO_OPTIONS_STOCK"),
+            "noOptionsExecute": null
```
I'm confused. I thought the only use case for this execution function was for Adobe Stock? But, it's null here. What's the deal?

iwehrman commented on src/js/actions/search/stock.js:
```diff
@@ -0,0 +1,123 @@
+            "type": "ADOBE_STOCK",
+            "getOptions": _adobeStockSearchOptions.bind(this),
+            "filters": Immutable.List.of("ADOBE_STOCK"),
+            "handleExecute": _confirmStockSearch.bind(this),
```
It looks like you're just using the standard execution function to handle default options. If that works then that's fine. Maybe there doesn't need to be a special execution function just for default options?

iwehrman replied:
This function also doesn't refer to `this`.

iwehrman commented on src/js/actions/search/stock.js:
```diff
@@ -0,0 +1,123 @@
+    var registerAdobeStock = function () {
+        var adobeStockPayload = {
+            "type": "ADOBE_STOCK",
+            "getOptions": _adobeStockSearchOptions.bind(this),
```
Again, this doesn't need to be bound because that function doesn't refer to `this`.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -400,6 +400,19 @@ define(function (require, exports, module) {
             this.refs.datalist.removeAutofillSuggestion();
+            
+            // If string, it means no items match user inputted search query
+            var filters = this.state.filter,
```
I'm still confused about why this is called `filters` (plural).

DivyaPrabhakar replied:
This is called `filters` because there can be more than one filter added at a time, i.e. "Layers" > "Text Layers"

iwehrman replied:
OK, but then why is the state called `filter`?

iwehrman replied:
The point is that the mismatch is confusing.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -400,6 +400,19 @@ define(function (require, exports, module) {
+                payload = searchStore.getPayloadFromFilter(lastFilter),
+                noOptionsStock = nls.localize("strings.SEARCH.NO_OPTIONS_STOCK");
+
+            if (payload && payload.noOptions().noOptionsString === noOptionsStock) {
```
You've created an explicit dependency on Stock. There should be no explicit dependency on any provider. Consider that if this ever gets moved to standard PS we will have to shuffle the set of providers somewhat, and we can't predict how exactly. So, there should be no explicit dependencies in any generic infrastructure on specific implementations.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -407,20 +420,24 @@ define(function (require, exports, module) {
          * Perform action based on ID
          *
          * @private
-         * @type {Datalist~onInputChange}
```
I think @shaoshing wants that annotation there to indicate that this function is used as the `onInputChange` handler to the child `Datalist` instance.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -407,20 +420,24 @@ define(function (require, exports, module) {
+                type;
+
             if (id === null) {
+                type = "";
```
If you're returning right after this, why bother initializing this variable?

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -407,20 +420,24 @@ define(function (require, exports, module) {
+                type = "";
                 this.props.dismissDialog();
                 return;
+            } else {
```
We prefer the following pattern:
```
if (badState) {
  cleanup();
  return;
}

// Normal case
...
```

Instead of:
```
if (badCase) {
  cleanup();
  return;
} else {
 // do stuff
}
```
We prefer the former because subjectively we find it easier to read by dismissing the exceptional cases first, and then proceeding with the main body at a lower indentation level.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -407,20 +420,24 @@ define(function (require, exports, module) {
                 this.props.dismissDialog();
                 return;
+            } else {
+                idSplit = id.split("-");
```
For future reference, I really hate how we have two separate data items encoded in a single concatenated string. This is a terrible programming idiom. This doesn't need to be fixed now, but it should be fixed...

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -458,13 +475,16 @@ define(function (require, exports, module) {
 
         /**
          * @private
-         * @type {Datalist~onKeyDown}
```
@shaoshing wants this too, and so on for the existing `@type` annotations.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -407,20 +420,24 @@ define(function (require, exports, module) {
+         * @param {string} id of the selected option
          */
         _handleChange: function (id) {
+            var filters = this.state.filter,
```
Also, all this is superfluous in case `id === null`.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -458,13 +475,16 @@ define(function (require, exports, module) {
-                case "Return":
                 case "Enter":
-                    if (selectedID === PLACEHOLDER_ID) {
+                    // The only time selectedID is falsey is when an Adobe Stock search
```
Presumably you are building infrastructure that is not specific to Adobe Stock, so this comment should not be specific to Adobe Stock.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -458,13 +475,16 @@ define(function (require, exports, module) {
          */
         _handleKeyDown: function (event, selectedID) {
             switch (event.key) {
-                case "Return":
```
Why remove this case?

chadrolfs commented:
One issue I noticed on this branch that I couldn't repro on master was that sometimes the first character gets lost when typing in search after the filter is applied.
cmd + f
arrow down to Adobe Stock, hit enter to set the filter
type Ki, hit delete twice to delete Ki, then type it again.  I end up with i

DivyaPrabhakar commented:
@chadrolfs Now this shouldn't be happening, can you verify?

iwehrman commented:
Clicking the "Search Adobe Stock" result (once the filter is applied) doesn't work. For every other search result, this works.

iwehrman commented:
The search result (but NOT the input placeholder text) should have an ellipsis at the end to give a hint that you're about to open up a new window:
![image](https://cloud.githubusercontent.com/assets/1075154/12521730/2fcd3768-c101-11e5-8beb-3b1962d0b567.png)

iwehrman commented:
I'm seeing this React warning whenever I perform a Stock search:
```
Warning: setState(...): Can only update a mounted or mounting component. This usually means you called setState() on an unmounted component. This is a no-op. Please check the code for the TextInput component.
```

DivyaPrabhakar commented:
@iwehrman Looking into the error now, the stock search filter with no user input should be fixed now. Opens to the homepage
@chadrolfs The error came back for me just FYI.

iwehrman commented on src/js/actions/search/stock.js:
```diff
@@ -0,0 +1,127 @@
+        nls = require("js/util/nls");
+
+    var SEARCH_STOCK_URL = "https://stock.adobe.com/search?k=",
+        ADOBE_STOCK_URL = "https://stock.adobe.com/at/";
```
This seems to be the Austrian home page for stock. I suggest just sending them to stock.adobe.com. But, this does raise the question of whether that's safe for all locales...

DivyaPrabhakar replied:
Lol. I'll change that. For some reason Adobe Stock has defaulted to German and I cannot change any of the settings and when I do it reverts. I have cleared my cache, changed settings in Creative Cloud, etc, nothing works. So I didn't realize that was the not the default.

DivyaPrabhakar commented:
@iwehrman The stock URL to the universal homepage has been changed.

DivyaPrabhakar commented:
@iwehrman Finally back to you -- the layer selection problem I had a brief freak out over is fixed, the warnings are not occurring any more, and the no options states should all be working too.

iwehrman commented:
I'm still seeing this warning when I choose the Stock search option:
```
Warning: setState(...): Can only update a mounted or mounting component. This usually means you 
called setState() on an unmounted component. This is a no-op. Please check the code for the TextInput 
component.

warning @ warning.js:45
getInternalInstanceReadyForUpdate @ ReactUpdateQueue.js:34
ReactUpdateQueue.enqueueSetState @ ReactUpdateQueue.js:181
ReactComponent.setState @ ReactComponent.js:65
(anonymous function) @ TextInput.jsx:345
tryCatcher @ bluebird.js:4978
Promise._settlePromiseFromHandler @ bluebird.js:3042
Promise._settlePromise @ bluebird.js:3099
...
```

I also get this new error when I press Escape to dismiss the empty search UI without typing anything else in:
```
Uncaught TypeError: Cannot read property 'close' of undefined

React.createClass._handleInputKeyDown @ Datalist.jsx:334
React.createClass._handleKeyDown @ TextInput.jsx:313
ReactErrorUtils.invokeGuardedCallback @ ReactErrorUtils.js:71
executeDispatch @ EventPluginUtils.js:79
executeDispatchesInOrder @ EventPluginUtils.js:102
executeDispatchesAndRelease @ EventPluginHub.js:43
executeDispatchesAndReleaseTopLevel @ EventPluginHub.js:54
forEachAccumulated @ forEachAccumulated.js:23
EventPluginHub.processEventQueue @ EventPluginHub.js:259
runEventQueueInBatch @ ReactEventEmitterMixin.js:18
ReactEventEmitterMixin.handleTopLevel @ ReactEventEmitterMixin.js:34
handleTopLevelWithoutPath @ ReactEventListener.js:93
handleTopLevelImpl @ ReactEventListener.js:73
Mixin.perform @ Transaction.js:136
ReactDefaultBatchingStrategy.batchedUpdates @ ReactDefaultBatchingStrategy.js:62
batchedUpdates @ ReactUpdates.js:94
ReactEventListener.dispatchEvent @ ReactEventListener.js:204
```

iwehrman commented:
I suspect the latter error will be easy to track down, but the former may be more difficult. I'm happy to take a look with you if you'd like!

iwehrman commented on src/js/stores/search.js:
```diff
@@ -271,38 +272,46 @@ define(function (require, exports, module) {
+                    // Get shortest unique paths
+                    var ancestors = collection.pluck(items, "pathInfo"),
+                        shortenedPaths = searchTypeInfo.shortenPaths ?
+                            pathUtil.getShortestUniquePaths(ancestors).toJS() : ancestors.toJS();
```
There's no need to call `toJS` here, which is slow. Instead of doing an array lookup below, you can not convert these immutable objects and call `.get(index)`. (I know you didn't create this bug, but you're in there now...)

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -458,20 +482,27 @@ define(function (require, exports, module) {
                 case "Enter":
-                    if (selectedID === PLACEHOLDER_ID) {
-                        this._handleDialogClick(event);
+                    // The only time selectedID is falsey is when an Adobe Stock search
```
This comment is surely out of date since you're not hard-coding anything about Adobe Stock in this component, right?

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -429,12 +443,16 @@ define(function (require, exports, module) {
+
+            var flux = this.getFlux(),
+                searchStore = flux.store("search"),
+                payload = searchStore.getPayloadFromFilter(lastFilter);
```
I guess this doesn't need to be changed here, but `payload` is not a very good word to use here because it's not descriptive at all. If I'm not familiar with this code, I'm scratching my head thinking "what the heck is a payload?". Ideally there would be a more descriptive term for whatever this function returns that makes it possible for me to keep reading without having to pause and go read that function to see roughly what the heck it does..

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -414,14 +426,16 @@ define(function (require, exports, module) {
                 return;
             }
 
             var filters = this.state.filter,
```
I think this was supposed to be renamed to `this.state.filters`?

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -171,8 +170,8 @@ define(function (require, exports, module) {
 
-            if (prevState.filter !== this.state.filter && this.refs.datalist) {
-                this._updateDatalistInput(this.state.filter);
+            if (prevState.filter !== this.state.filters && this.refs.datalist) {
```
This should also be `.filters`.

iwehrman commented on src/js/actions/search/stock.js:
```diff
@@ -0,0 +1,127 @@
+
+    /**
+     * Register Adobe Stock info for search
+     */
```
Missing `@return {Promise}`.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -445,6 +463,12 @@ define(function (require, exports, module) {
+                if (payload && payload.noOptionsDefault) {
+                    this.props.executeOption(searchTerm, payload.noOptionsDefault().noOptionsExecute);
+                } else {
+                    return { dontCloseDialog: true };
```
This function is also missing `@return {{dontCloseDialog: boolean}=}`.

DivyaPrabhakar replied:
I am assuming you mean a `@return {{dontCloseDialog: boolean}=}` at line 418 which has now been added but this comment won't be outdated because the change was not made to line 470.

iwehrman commented:
I'm still seeing the same `setState`-after-component-unmounted warning when I click the `Search Adobe Stock...` search result. It's still in the same place in `_beginEdit.`

iwehrman commented:
Back to @DivyaPrabhakar. Getting close!

DivyaPrabhakar commented:
@iwehrman and @ktaki: This PR has the fixes for the issues #3529 and #3524

shaoshing commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -415,14 +426,17 @@ define(function (require, exports, module) {
          *
          * @private
          * @type {Datalist~onInputChange}
+         * @param {string} id of the selected option
```
Missing the `searchTerm` JSDoc

shaoshing commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -230,9 +230,20 @@ define(function (require, exports, module) {
-                        return idString.includes(word.toLowerCase()) ? "" : word;
+                        if (idString.includes(word.toLowerCase())) {
+                            partialMatch = true;
+                            // return "";
```
Should this line be removed or uncommented?

ktaki commented:
@DivyaPrabhakar: Somehow this can be reproduced with Windows platform only. Open the search bar. Type any single character. The error below happens.

![searchbarerror](https://cloud.githubusercontent.com/assets/2029597/12659339/26e76120-c5c3-11e5-9846-ef2f81ac7e15.png)

DivyaPrabhakar commented:
The current state of this branch is that it addresses all the bugs referenced in this PR as well as all the architecture concerns. @iwehrman -- sending to you to look over. 

Summary of the current situation: 
Search on `master` is currently broken -- keyup and keydown skips options in the datalist, pressing enter on a datalist option is broken. This is because of the change in event handling -- for every key press, on Windows we were getting `onkeydown` and `oninput` events; on Mac we were only getting `oninput` events. Therefore, we were seeing errors thrown on Mac that we were not seeing on Windows. That new event that is getting registered is causing the problems on datalist. If those get solved, the changes in this PR should ideally integrate perfectly within DS; only time will tell...

ktaki commented:
In the search dialog, select Adobe Stock. Enter a word to search. Hit return.

![screen shot 2016-02-17 at 11 06 51 am](https://cloud.githubusercontent.com/assets/2029597/13121308/c01944fa-d566-11e5-9186-6d650b23d730.png)

iwehrman commented on src/js/jsx/Search.jsx:
```diff
@@ -53,14 +53,19 @@ define(function (require, exports, module) {
          * 
          * @private
          * @param {string} itemID ID of selected option
+         * @param {string=} noOptionsFn function to execute if there are no options
```
`noOptionsFn` does not look like a string to me!

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -216,7 +215,8 @@ define(function (require, exports, module) {
         _updateDatalistInput: function (id) {
             if (id) {
-                var currFilter = this.refs.datalist.getInputValue().split(" "),
+                var partialMatch,
```
Style nit: in var declaration blocks, put the declarations with definitions first.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -230,9 +230,19 @@ define(function (require, exports, module) {
                     nextFilterMap = _.map(currFilter, function (word) {
-                        return idString.includes(word.toLowerCase()) ? "" : word;
+                        if (idString.includes(word.toLowerCase())) {
+                            partialMatch = true;
```
Map functions should always return a value. It's weird that this function only returns a value sometimes. If this is intentional, you should explicitly return `null` instead of having no return.

iwehrman replied:
(Or the empty string, or whatever the right explicit value should be.)

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -216,7 +215,8 @@ define(function (require, exports, module) {
          */
         _updateDatalistInput: function (id) {
```
It's super, super confusing that this variable is called `id` when it is actually an array of `id`s.

iwehrman commented on src/nls/en/strings.json:
```diff
@@ -468,7 +468,9 @@
         "PLACEHOLDER": "Search, open & select",
         "PLACEHOLDER_INITIALIZING": "Initializing…",
         "PLACEHOLDER_FILTER": "Search ",
+        "NO_OPTIONS_STOCK": "Search Adobe Stock...",
```
This should be an ellipsis, not three periods.

iwehrman commented on src/js/stores/search.js:
```diff
@@ -271,38 +272,46 @@ define(function (require, exports, module) {
+                    itemMap = items.map(function (item, index) {
+                        var newPathInfo = shortenedPaths.get(index) || "",
+                            name = item.name,
+                            style,
```
Style nit: reorder declarations.

iwehrman commented on src/js/stores/search.js:
```diff
@@ -271,38 +272,46 @@ define(function (require, exports, module) {
+                            style: style,
+                            type: "item"
+                        };
+                    }.bind(this));
```
For collections functions that accept a receiver argument, that is preferred over an explicit bind.

iwehrman commented on src/js/jsx/shared/Datalist.jsx:
```diff
@@ -701,25 +735,26 @@ define(function (require, exports, module) {
                 );
             }
-
+           
```
Bad whitespace change.

iwehrman commented on src/js/jsx/shared/Datalist.jsx:
```diff
@@ -603,7 +626,18 @@ define(function (require, exports, module) {
+            }
+            
+            if (currentOptions && currentOptions.size === 1) {
+                id = currentOptions.toJS()[0].id;
```
Calling `.toJS()` is very slow. It would be better to not do so if possible. It seems like you could just do `currentOptions.first().id` instead.

iwehrman replied:
Also it seems like this whole if block could be moved inside the previous one?

iwehrman replied:
Or, even better:
```javascript
if (!this.refs.select) {
  return;
}
```

iwehrman commented on src/js/jsx/shared/Datalist.jsx:
```diff
@@ -323,10 +323,27 @@ define(function (require, exports, module) {
 
-            var cbOptions = this.props.onKeyDown(event, this.state.lastHighlightedID),
-                options = _.merge({ preventListDefault: false }, cbOptions);
+            // All the potential options for the curret user input 
```
Typo: `curret`

iwehrman commented on src/js/jsx/shared/Datalist.jsx:
```diff
@@ -323,10 +323,27 @@ define(function (require, exports, module) {
-            var cbOptions = this.props.onKeyDown(event, this.state.lastHighlightedID),
-                options = _.merge({ preventListDefault: false }, cbOptions);
+            // All the potential options for the curret user input 
+            var optionsList = select ? select.props.options : null,
```
It seems like this initialization could be delayed?

iwehrman commented on src/js/jsx/shared/Datalist.jsx:
```diff
@@ -340,9 +357,15 @@ define(function (require, exports, module) {
-                    select.close(event, "apply");
-                    if (dialog && dialog.isOpen()) {
-                        dialog.toggle(event);
+                    var noOptionsExecute = optionsList ? optionsList.toJS()[0].handleExecute : null;
```
Again, if possible, please avoid calling `.toJS()`.

iwehrman replied:
I don't think that `Datallist` should be doing anything that involves `handleExecute`, which is only relevant to `Search` AFAIK. Can you help me understand what this is used for so we can find a way to move this functionality out of `Datalist`?

iwehrman commented on src/js/stores/search.js:
```diff
@@ -271,38 +272,46 @@ define(function (require, exports, module) {
+                    // Get shortest unique paths
+                    var ancestors = collection.pluck(items, "pathInfo"),
+                        shortenedPaths = searchTypeInfo.shortenPaths ?
+                                pathUtil.getShortestUniquePaths(ancestors) : ancestors;
```
Looks like you jumped a tab stop here.

iwehrman commented:
@DivyaPrabhakar: to fix the "Can only update a mounted component" problem:

1) Save the promise from `acquireFocus()` in object state, with something like:
```javascript
this._acquireFocusPromise = this.acquireFocus()
  .bind(this)
  .then(/* whatever */)
  .finally(function () {
    this._acquireFocusPromise = null;
  });
```

2) In `componentWillUnmount`, do something like: 
```javascript
if (this._acquireFocusPromise && /* status is "not yet resolved" */) {
  this._acquireFocusPromise.cancel();
}
```

If necessary, repeat for `releaseFocus`. 

CC @volfied: is this something that we could integrate into `Focusable`?

ktaki commented:
@DivyaPrabhakar: Tested the latest branch.

#3610 Autocompletion by tab key is working.
#3529 Still reproducible
#3524 TypeError is gone, but `Warning: setState(...):` is still happening.

The result of the first search (Limit search to...) is not listed in the dialog. For example, typing Layers and hit return does not list layers. Is this what you were referring to as "broken"?

iwehrman commented:
I think it's OK if #3529 is not fixed in this PR. The `setState` problem will be fixed in short order. The "result of first search" sounds like a blocker though.

volfied commented:
@iwehrman We can improve Focusable mixin to make sure the focus functions are only called while component is mounted. However, we wouldn't be able to have 1) where something happens in the then handler of the owner component, and then the promise pointer is set to null.

DivyaPrabhakar commented:
Updated the PR w.r.t. to the comments made. @ktaki #3529 should be fixed with these changes now. @iwehrman I removed the search specific checks from datalist that you asked for clarification about and moved it to searchbar.jsx (you can see what I am talking about on in searchbar.jsx, line 505-509 on this branch)

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -229,10 +227,23 @@ define(function (require, exports, module) {
 
                         return toRemove ? toRemove.replace(" ", "") : idWord;
                     }, this).join("").toLowerCase(),
+                    partialMatch,
```
Again, declarations without a definition should be moved to the end of var declaration blocks for stylistic consistency with the rest of the code.

DivyaPrabhakar replied:
This variable is used by the subsequent variable declarations, therefore must be present here. sorry for the ugly.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -416,15 +427,19 @@ define(function (require, exports, module) {
-        _handleChange: function (id) {
-            if (id === null) {
+        _handleChange: function (id, searchTerm) {
+            if (!id) {
```
Given that `id` is a non-optional parameter, this will only match if `id === ""`. Is that intentional?

DivyaPrabhakar replied:
I confirmed that this `id` is in fact optional (there was a mistake in the JSDoc), therefore this check makes more sense now and I have kept it in.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -416,15 +427,19 @@ define(function (require, exports, module) {
 
-            var filters = this.state.filter,
+            var filters = this.state.filters,
+                lastFilter = filters ? filters[filters.length - 1] : id,
```
If filters is an empty array, this will throw.

iwehrman replied:
Er, actually, it will just be undefined, because `filters.length - 1` will be -1, and `[-1] === undefined`. Anyway, Is this what you want?

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -462,20 +487,30 @@ define(function (require, exports, module) {
          * @type {Datalist~onKeyDown}
+         * @param {SyntheticEvent} event
+         * @param {string} selectedID the id of the selected item in the datalist
+         * @param {list=} optionsList
```
This is not a valid JSDoc type expression.

iwehrman commented on src/js/jsx/shared/Datalist.jsx:
```diff
@@ -586,7 +601,7 @@ define(function (require, exports, module) {
         /**
          * Sets autofill suggestion to empty string and re-renders
-         *
+         * 
```
Bad whitespace change.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -416,15 +429,19 @@ define(function (require, exports, module) {
          *
          * @private
-         * @type {Datalist~onChange}
+         * @type {Datalist~onInputChange}
```
There is no longer an `onInputChange` prop to `Datalist`.

chadrolfs commented:
Tested this branch and all the bugs look fixed (tested Mac and Win).  Also, everything functions as expected and explained above.  From a user/functional standpoint, I think it's ready to merge.  @ktaki wanna have a look?

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -416,15 +429,19 @@ define(function (require, exports, module) {
-        _handleChange: function (id) {
-            if (id === null) {
+        _handleChange: function (id, searchTerm) {
+            if (id === null || !id) {
```
This is redundant, since if `id === null` then `!id` would suffice. Also the param type annotation says that string is required, so that would imply that it can't be null. Do you know if `id` is actually ever null?

DivyaPrabhakar replied:
This can be null (see: datalist.jsx `_handleSelectClick`), so I will change the parameter to optional and only have the `!id` check.

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -432,13 +449,21 @@ define(function (require, exports, module) {
 
             // Seeing if filters are active or not for analytics
-            if (filters.length === 0) {
+            if (filters && filters.length === 0) {
```
Is `filters` guaranteed to be an array? If so, the first conjunct is unnecessary. If not, you probably want `!filter || filters.length === 0` instead?

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -432,13 +449,21 @@ define(function (require, exports, module) {
+            var flux = this.getFlux(),
+                searchStore = flux.store("search"),
+                // This payload will be used to dictate whether given the applied filter (0 or more), if 
+                // we want to take any action in the case the user input returns in no search results
```
I think you want `input returns no search results.` ?

iwehrman commented on src/js/jsx/search/SearchBar.jsx:
```diff
@@ -432,13 +449,21 @@ define(function (require, exports, module) {
+
+            var flux = this.getFlux(),
+                searchStore = flux.store("search"),
+                // This payload will be used to dictate whether given the applied filter (0 or more), if 
```
`applied filters` ?

ktaki commented:
Works much better :+1: Agree with Chad. Ready to merge. Will pound on it more on master.

iwehrman commented:
Let's do it!
![img](https://s3.amazonaws.com/f.cl.ly/items/3T0j3v301t300s0c461A/thumbs-up-montage.gif?v=4f4db623)
